### PR TITLE
feat: Import 2PC — CommitImport/AbortImport with auto_commit and WAL broadcast

### PR DESCRIPTION
**Pull request by bigsheeper**
## Summary

Implements Two-Phase Commit (2PC) for Import in primary/secondary replication clusters. Data stays invisible (`is_importing=true`) until an explicit commit signal is delivered via WAL, ensuring primary and secondary clusters reach the same visible state at the same logical position.

- **New proto**: `CommitImport=44`, `RollbackImport=45` WAL message types; `Uncommitted=8`, `Committing=9` `ImportJobState` values; `CommitImport`, `AbortImport`, `HandleCommitVchannel` RPCs on DataCoord; `committed_vchannels` + `auto_commit` fields on `ImportJob`
- **WAL broadcast**: DataCoord broadcasts `CommitImportMessage`/`RollbackImportMessage` to all vchannels via DDL broadcast; CDC replicates to secondary clusters verbatim
- **DDL ack callbacks**: `commitImportV2AckCallback` CAS `Uncommitted→Committing`; `rollbackImportV2AckCallback` CAS `*→Failed` + segment drop
- **WAL flusher**: `wal_flusher.dispatch()` intercepts `CommitImportMessage` per-vchannel, calls `wbMgr.FlushChannel` + `DataCoord.HandleCommitVchannel`; no-op handler for `RollbackImportMessage`
- **ImportChecker**: new `Uncommitted` case (auto-commits when `auto_commit=true`); new `Committing` case (transitions to `Completed` when all vchannels confirmed)
- **auto_commit option**: default `true` (backward compatible); `false` lets replication platform control commit timing
- **RESTful API**: `POST /v2/vectordb/jobs/import/commit` and `POST /v2/vectordb/jobs/import/abort`; `GetImportProgress` surfaces `Uncommitted` and `Committing` states

**Out of scope**: `commit_timestamp` propagation (handled in companion PR)

## Test Plan

- [ ] Unit tests for `IsAutoCommit` helper
- [ ] Unit tests for `HandleCommitVchannel` idempotency
- [ ] Unit tests for `CommitImport`/`AbortImport` RPC handlers (state validation, mutex, broadcast)
- [ ] Unit tests for DDL ack callbacks (CAS races for commit/abort)
- [ ] Unit tests for `ImportChecker` `Uncommitted`/`Committing` cases
- [ ] Unit tests for WAL flusher dispatch (CommitImport/RollbackImport cases)
- [ ] Unit tests for RESTful commit/abort handlers (valid jobId, invalid jobId error path)
- [ ] `GetImportProgress` surfaces `Uncommitted` and `Committing` states
- [ ] E2E: non-replication cluster with `auto_commit=true` (default) behaves identically to pre-2PC

issue: #48525
design doc: https://github.com/milvus-io/milvus-design-docs/pull/29

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/datacoord/ddl_callbacks_import.go [hunk 1]
```diff
@@ -187,3 +188,103 @@ func (s *Server) broadcastImport(ctx context.Context,
...
+			zap.Int64("jobID", jobID))
+		return nil
+	}
+
+	// Drop all import segments for this job.
+	return c.dropImportJobSegments(ctx, jobID)
+}
+
+// dropImportJobSegments marks all segments belonging to the given import job as Dropped.
+func (c *DDLCallbacks) dropImportJobSegments(ctx context.Context, jobID int64) error {
```
Comment (bigsheeper) [inline internal/datacoord/ddl_callbacks_import.go]: segments will be dropped in import inspector, don't drop here

File: internal/datacoord/ddl_callbacks_import_test.go [hunk 2]
```diff
@@ -822,3 +822,19 @@ func TestImportFlowIntegration(t *testing.T) {
 		t.Log("Import flow documented successfully")
 	})
 }
+
+func TestCommitImportCallback_UncommittedToCommitting(t *testing.T) {
+	t.Skip("environment: requires full server setup")
```
Comment (bigsheeper) [inline internal/datacoord/ddl_callbacks_import_test.go]: fix them

File: internal/datacoord/import_checker.go [hunk 3]
```diff
@@ -266,8 +276,15 @@ func (c *importChecker) checkPreImportingJob(job ImportJob) {
 	}
 
 	if totalRows == 0 {
-		log.Info("no data to import, skip the subsequent stages, just update job state to Completed")
-		updateJobState(internalpb.ImportJobState_Completed)
+		if job.GetAutoCommit() {
+			// backward-compatible: auto-commit jobs skip Uncommitted directly to Completed
```
Comment (bigsheeper) [inline internal/datacoord/import_checker.go]: don't describe auto-commit as a backward-compatible

File: internal/datacoord/import_checker.go [hunk 4]
```diff
@@ -444,6 +461,49 @@ func (c *importChecker) checkIndexBuildingJob(job ImportJob) {
...
+// If auto_commit=false, it waits for an explicit CommitImport RPC from the platform.
+func (c *importChecker) checkUncommittedJob(job ImportJob) {
+	log := log.With(zap.Int64("jobID", job.GetJobID()))
+	if !job.GetAutoCommit() {
+		// Wait for explicit CommitImport from the replication platform.
+		return
+	}
+	// auto_commit=true: trigger commit by broadcasting the WAL message.
+	if c.commitImportFn == nil {
+		log.Error("commitImportFn is nil but auto_commit=true; job will stall in Uncommitted state",
```
Comment (bigsheeper) [inline internal/datacoord/import_checker.go]: should be unexpected? can we just panic here?

File: internal/datacoord/import_checker.go [hunk 5]
```diff
@@ -444,6 +461,49 @@ func (c *importChecker) checkIndexBuildingJob(job ImportJob) {
...
+			zap.Int64("jobID", job.GetJobID()))
+		return
+	}
+	if err := c.commitImportFn(c.ctx, job); err != nil {
+		log.Warn("auto-commit import failed", zap.Error(err))
+	}
```
Comment (bigsheeper) [inline internal/datacoord/import_checker.go:479]: find the right place to add the timecost log for uncommited -> committing stage

File: internal/datacoord/import_job.go [hunk 6]
```diff
@@ -98,13 +98,29 @@ func UpdateJobCompleteTime(completeTime string) UpdateJobAction {
 	}
 }
 
+// UpdateJobStateWithCAS returns an UpdateJobAction that transitions from→to only
+// if the current state matches 'from'. If the state has already changed (race:
+// abort won before commit, or commit won before abort), this is a no-op.
+// It delegates to UpdateJobState so all terminal-state side effects (CleanupTs,
+// RequestedDiskSize reset, etc.) are applied correctly.
+func UpdateJobStateWithCAS(from, to internalpb.ImportJobState) UpdateJobAction {
```
Comment (bigsheeper) [inline internal/datacoord/import_job.go]: does it ensure the concurrent safety?

File: internal/datacoord/import_meta.go [hunk 7]
```diff
@@ -335,3 +338,30 @@ func (m *importMeta) TaskStatsJSON(ctx context.Context) string {
 	}
 	return string(ret)
 }
+
+func (m *importMeta) HandleCommitVchannel(ctx context.Context, jobID int64, vchannel string, callback func() error) error {
```
Comment (bigsheeper) [inline internal/datacoord/import_meta.go:342]: Just question, why not wrap to a UpdateJobAction?

File: internal/datacoord/import_util.go [hunk 8]
```diff
@@ -600,6 +600,14 @@ func GetJobProgress(ctx context.Context, jobID int64,
 		_, totalRows := getImportRowsInfo(ctx, jobID, importMeta, meta)
 		return 10 + 30 + 30 + 10 + int64(progress*10), internalpb.ImportJobState_Importing, totalRows, totalRows, ""
 
+	case internalpb.ImportJobState_Uncommitted:
+		_, totalRows := getImportRowsInfo(ctx, jobID, importMeta, meta)
+		return 100, internalpb.ImportJobState_Uncommitted, totalRows, totalRows, ""
```
Comment (bigsheeper) [inline internal/datacoord/import_util.go]: maybe 99 is better?

File: internal/datacoord/import_util.go [hunk 9]
```diff
@@ -600,6 +600,14 @@ func GetJobProgress(ctx context.Context, jobID int64,
...
+
+	case internalpb.ImportJobState_Committing:
+		_, totalRows := getImportRowsInfo(ctx, jobID, importMeta, meta)
+		return 100, internalpb.ImportJobState_Committing, totalRows, totalRows, ""
```
Comment (bigsheeper) [inline internal/datacoord/import_util.go]: maybe 99 is better?

File: internal/datacoord/services.go [hunk 10]
```diff
@@ -2570,3 +2572,213 @@ func (s *Server) ListRefreshExternalCollectionJobs(ctx context.Context, req *dat
 		Jobs:   jobs,
 	}, nil
 }
+
+// getOrCreateImportJobMu returns a per-job mutex for serializing concurrent commit/abort on the same job.
+func (s *Server) getOrCreateImportJobMu(jobID int64) *sync.Mutex {
+	actual, _ := s.importJobMu.LoadOrStore(jobID, &sync.Mutex{})
```
Comment (bigsheeper) [inline internal/datacoord/services.go]: use keylock？

File: internal/datacoord/services.go [hunk 11]
```diff
@@ -2570,3 +2572,213 @@ func (s *Server) ListRefreshExternalCollectionJobs(ctx context.Context, req *dat
...
+			fmt.Sprintf("job %d is in state %s, expected Uncommitted", jobID, state))), nil
+	}
+
+	// Acquire per-job mutex to serialize concurrent commit/abort.
+	mu := s.getOrCreateImportJobMu(jobID)
+	mu.Lock()
+	defer mu.Unlock()
+
+	// Re-check state after acquiring mutex (another goroutine may have changed it).
+	job = s.importMeta.GetJob(ctx, jobID)
```
Comment (bigsheeper) [inline internal/datacoord/services.go:2914]: check if is auto-commit? if is, report error.

File: internal/datacoord/services.go [hunk 12]
```diff
@@ -2570,3 +2572,213 @@ func (s *Server) ListRefreshExternalCollectionJobs(ctx context.Context, req *dat
...
+			fmt.Sprintf("job %d is in terminal/committed state %s, abort not allowed", jobID, state))), nil
+	}
+
+	// Acquire per-job mutex to serialize concurrent commit/abort.
+	mu := s.getOrCreateImportJobMu(jobID)
+	mu.Lock()
+	defer mu.Unlock()
+
+	// Re-check state after acquiring mutex.
+	job = s.importMeta.GetJob(ctx, jobID)
```
Comment (bigsheeper) [inline internal/datacoord/services.go]: reduce dup codes

File: internal/datacoord/import_meta.go [hunk 13]
```diff
@@ -335,3 +338,30 @@ func (m *importMeta) TaskStatsJSON(ctx context.Context) string {
...
+	// This ensures callback is only called once even on retry — if save fails,
+	// the caller retries and the callback is not invoked until save succeeds.
+	updatedJob := job.Clone()
+	updatedJob.(*importJob).ImportJob.CommittedVchannels = append(updatedJob.GetCommittedVchannels(), vchannel)
+	if err := m.catalog.SaveImportJob(ctx, updatedJob.(*importJob).ImportJob); err != nil {
+		return err
+	}
+	m.jobs[jobID] = updatedJob
+	// Invoke caller-supplied callback after successful persist
+	return callback()
```
Comment (bigsheeper) [inline internal/datacoord/import_meta.go]: Not an atomic op, save import job + callback should be a transaction?

File: pkg/proto/data_coord.proto [hunk 14]
```diff
@@ -1253,6 +1258,8 @@ message ImportJob {
   string create_time = 16;
   repeated string ready_vchannels = 17;
```
Comment (bigsheeper) [inline pkg/proto/data_coord.proto:1282]: is ready_vchannels could deprecated? if so, mark as deprecated

File: pkg/proto/messages.proto [hunk 15]
```diff
@@ -627,6 +629,26 @@ message RefreshExternalCollectionMessageHeader {
 // Empty - all info is in header.
 message RefreshExternalCollectionMessageBody {}
 
+// CommitImportMessageHeader is the header of commit import message.
+message CommitImportMessageHeader {
+    int64 collection_id = 1;
```
Comment (bigsheeper) [inline pkg/proto/messages.proto:650]: do we use this field?

File: pkg/proto/messages.proto [hunk 16]
```diff
@@ -627,6 +629,26 @@ message RefreshExternalCollectionMessageHeader {
...
+    int64 job_id        = 2;
+}
+
+// CommitImportMessageBody is the body of commit import message.
+// Empty - all info is in header.
+message CommitImportMessageBody {}
+
+// RollbackImportMessageHeader is the header of rollback import message.
+message RollbackImportMessageHeader {
+    int64 collection_id = 1;
```
Comment (bigsheeper) [inline pkg/proto/messages.proto:660]: do we use this field?

File: internal/datacoord/ddl_callbacks_import_test.go [hunk 17]
```diff
@@ -822,3 +827,139 @@ func TestImportFlowIntegration(t *testing.T) {
 		t.Log("Import flow documented successfully")
 	})
 }
+
```
Comment (bigsheeper) [inline internal/datacoord/ddl_callbacks_import_test.go:831]: Add concurrent ut

File: internal/datacoord/services.go [hunk 18]
```diff
@@ -2570,3 +2571,198 @@ func (s *Server) ListRefreshExternalCollectionJobs(ctx context.Context, req *dat
...
+	vchannel := req.GetVchannel()
+
+	// Pre-fetch segment IDs for this job+vchannel BEFORE calling HandleCommitVchannel.
+	// The callback must not access importMeta because HandleCommitVchannel holds m.mu (write lock);
+	// calling GetTaskBy inside the callback would attempt to re-acquire m.mu (read lock) → deadlock.
+	segIDs := s.getImportSegmentIDsByVchannel(ctx, jobID, vchannel)
+
+	err := s.importMeta.HandleCommitVchannel(ctx, jobID, vchannel, func() error {
+		// Only access s.meta (segment meta) here, NOT s.importMeta.
+		for _, segID := range segIDs {
```
Comment (bigsheeper) [inline internal/datacoord/services.go:2993]: Can we try to batch save?

File: internal/datacoord/services_test.go [hunk 19]
```diff
@@ -3518,3 +3518,18 @@ func TestServer_CreateExternalCollection_NoOp(t *testing.T) {
 	assert.NotNil(t, resp)
 	assert.False(t, merr.Ok(resp.GetStatus()))
 }
+
+func TestCommitImport_HappyPath(t *testing.T) {
+	// TODO: implement full test once DataCoord Server test infrastructure is wired up.
```
Comment (bigsheeper) [inline internal/datacoord/services_test.go]: Implement these tests.

File: internal/streamingnode/server/flusher/flusherimpl/wal_flusher.go [hunk 20]
```diff
@@ -278,6 +281,45 @@ func (impl *WALFlusherImpl) dispatch(msg message.ImmutableMessage) (err error) {
...
+		}
+		vchannel := msg.VChannel()
+		jobID := commitMsg.Header().GetJobId()
+
+		// Trigger async DML flush to ensure DML before this commit is flushed.
+		// FlushChannel failure is non-fatal: log warn and continue.
+		if err := resource.Resource().WriteBufferManager().
+			FlushChannel(context.Background(), vchannel, msg.TimeTick()); err != nil {
+			impl.logger.Warn("FlushChannel on CommitImport failed (non-fatal)",
+				zap.String("vchannel", vchannel), zap.Int64("jobID", jobID), zap.Error(err))
```
Comment (bigsheeper) [inline internal/streamingnode/server/flusher/flusherimpl/wal_flusher.go]: not just log, don't ignore the error

File: internal/streamingnode/server/flusher/flusherimpl/wal_flusher_test.go [hunk 21]
```diff
@@ -153,6 +153,12 @@ func newMockMixcoord(t *testing.T, maybe bool) *mocks.MockMixCoordClient {
 	return mixcoord
 }
 
+// TestDispatch_CommitImportMessage: verifies that dispatch() calls FlushChannel and HandleCommitVchannel.
```
Comment (bigsheeper) [inline internal/streamingnode/server/flusher/flusherimpl/wal_flusher_test.go]: Implement tests.

File: internal/datacoord/import_checker.go [hunk 22]
```diff
@@ -442,6 +459,47 @@ func (c *importChecker) checkIndexBuildingJob(job ImportJob) {
 	log.Info("import job all completed", zap.Duration("jobTimeCost/total", totalDuration))
 }
 
+// checkUncommittedJob handles jobs in the Uncommitted state.
+// If auto_commit=true, it triggers a commit via broadcastCommitImportMessage.
+// If auto_commit=false, it waits for an explicit CommitImport RPC from the platform.
+func (c *importChecker) checkUncommittedJob(job ImportJob) {
```
Comment (czs007) [inline internal/datacoord/import_checker.go:462]: **Blocker: the 2PC flow is never entered by non-empty imports.**

For a job that actually carries data the state machine is `Pending → PreImporting → Importing → Sorting → IndexBuilding → Completed`. `checkIndexBuildingJob` (line 450) unconditionally calls `unsetSegmentImporting` and then `UpdateJobState(Completed)` — there is no `auto_commit` gate and no `IndexBuilding → Uncommitted` transition.

The only place in the codebase that writes `Uncommitted` is `checkPreImportingJob` at import_checker.go:286, i.e. the `totalRows == 0` (empty import) branch. So `checkUncommittedJob`/`checkCommittingJob` and the entire CommitImport/AbortImport/WAL-fence machinery are only reachable for zero-row imports, which carry no data. For any real import the segments are made visible and the job completes exactly as before this PR — the "data stays invisible until an explicit commit" guarantee is not implemented.

This also contradicts the design doc (milvus-io/milvus-design-docs#29): the MEP's state machine explicitly includes `IndexBuilding → Uncommitted → Committing → Completed`, and its "Segment Visibility" section requires `is_importing` to stay `true` through `Uncommitted` and be cleared by `HandleCommitVchannel`, not by `checkIndexBuildingJob`.

Suggested fix: after index build completes, branch on `auto_commit` to `Uncommitted` vs `Completed`, and do not call `unsetSegmentImporting` before the commit fence — let the `HandleCommitVchannel` callback clear `is_importing` per vchannel. The empty-import fast path can stay.

File: internal/datacoord/services.go [hunk 23]
```diff
@@ -2821,3 +2822,194 @@ func (s *Server) ListRefreshExternalCollectionJobs(ctx context.Context, req *dat
...
+	}
+	defer broadcaster.Close()
+
+	msg := message.NewCommitImportMessageBuilderV2().
+		WithHeader(&message.CommitImportMessageHeader{
+			CollectionId: job.GetCollectionID(),
+			JobId:        job.GetJobID(),
+		}).
+		WithBody(&messagespb.CommitImportMessageBody{}).
+		WithBroadcast([]string{streaming.WAL().ControlChannel()}).
```
Comment (czs007) [inline internal/datacoord/services.go]: **Blocker: CommitImport/RollbackImport are broadcast to the control channel only, so the WAL flusher drops them before the per-vchannel handler runs.**

`broadcastCommitImportMessage` (and `broadcastRollbackImportMessage`, line 2862) use `WithBroadcast([]string{streaming.WAL().ControlChannel()})`. `WithBroadcast` does not set the pchannel-level property (only `WithClusterLevelBroadcast` does), so `IsPChannelLevel()` is false.

In wal_flusher.go `dispatch`, before `case message.MessageTypeCommitImport`, there is:

```
if funcutil.IsControlChannel(msg.VChannel()) && !msg.IsPChannelLevel() {
    return nil
}
```

A control-channel, non-pchannel-level message returns here, so the `CommitImport` case is never reached. `FlushChannel` + `HandleCommitVchannel` never run per data vchannel, `committed_vchannels` stays empty, and `checkCommittingJob`'s `len(committedVchannels) < len(Vchannels)` is always true for any collection with ≥1 data vchannel → the job is stuck in `Committing` forever.

The existing V1 `broadcastImport` correctly uses `WithBroadcast(vchannels)` (the job's data vchannels). The design doc (milvus-io/milvus-design-docs#29) also states Phase 1 should "Broadcast CommitImportMessage … to all vchannels (WAL)", and its "Rejected Alternatives" section specifically rejects doing the visibility flip without the per-vchannel roundtrip.

Suggested fix: broadcast to the job's data vchannels (as `broadcastImport` does); the control-channel copy is only enough to drive the DDL ack callback (`Uncommitted → Committing`). Also update `TestDispatch_CommitImportMessage`, which currently builds the message on a data vchannel and therefore does not exercise the real control-channel routing.

File: internal/datacoord/ddl_callbacks_import.go [hunk 24]
```diff
@@ -187,3 +188,69 @@ func (s *Server) broadcastImport(ctx context.Context,
...
+	registry.RegisterImportV1AckCallback(c.importV1AckCallback)
+	registry.RegisterCommitImportV2AckCallback(c.commitImportV2AckCallback)
+	registry.RegisterRollbackImportV2AckCallback(c.rollbackImportV2AckCallback)
+}
+
+// commitImportV2AckCallback handles the ack callback for CommitImport WAL message.
+// It transitions the import job from Uncommitted → Committing state.
+// Concurrency safety is guaranteed by the broadcaster framework's resource key lock
+// (exclusive collection-level lock), so no CAS is needed here.
+func (c *DDLCallbacks) commitImportV2AckCallback(ctx context.Context, result message.BroadcastResultCommitImportMessageV2) error {
```
Comment (czs007) [inline internal/datacoord/ddl_callbacks_import.go:202]: (non-blocking) `commitImportV2AckCallback`/`rollbackImportV2AckCallback` are a non-atomic read-check-write (GetJob → check state → UpdateJob). The comment says concurrency is guaranteed by the broadcaster's exclusive collection-level resource key lock. That reasoning holds (both messages are ExclusiveRequired and share the same `NewExclusiveCollectionNameResourceKey`), and the per-callback state guards give last-writer-wins — so it is acceptable. Two follow-ups: (1) CAS was deliberately removed in 9d47db2, which diverges from design doc milvus-io/milvus-design-docs#29, whose "Race Safety" table still specifies "CAS: Uncommitted → Committing". Please either update the MEP with the resource-key-lock argument or restore the CAS so doc and code agree. (2) There is no concurrency test exercising a commit/abort race; this invariant has no regression protection (the self-review note "Add concurrent ut" was marked resolved but the test was not added).

File: internal/datacoord/ddl_callbacks_import.go [hunk 25]
```diff
@@ -187,3 +188,69 @@ func (s *Server) broadcastImport(ctx context.Context,
...
+	state := job.GetState()
+	if state == internalpb.ImportJobState_Committing ||
+		state == internalpb.ImportJobState_Completed ||
+		state == internalpb.ImportJobState_Failed {
+		log.Ctx(ctx).Info("RollbackImport: job already in terminal/committed state, no-op",
+			zap.Int64("jobID", jobID), zap.String("state", state.String()))
+		return nil
+	}
+
+	return c.importMeta.UpdateJob(ctx, jobID, UpdateJobState(internalpb.ImportJobState_Failed))
```
Comment (czs007) [inline internal/datacoord/ddl_callbacks_import.go:255]: (non-blocking) Per design doc milvus-io/milvus-design-docs#29, the AbortImport ack should do four things: transition to `Failed`, set `RequestedDiskSize = 0` (release disk quota), set `CleanupTs = now + retention` (make GC-eligible), and mark the job's import segments `Dropped`. This callback only does `UpdateJobState(Failed)` and relies on the import inspector's `processFailed` for segment drop. Please confirm that disk quota is actually released and GC eligibility is set for an aborted job (the existing Failed path may not set `RequestedDiskSize=0`/`CleanupTs`), otherwise aborted-job resources may linger.

File: internal/datacoord/import_checker.go [hunk 26]
```diff
@@ -442,6 +459,47 @@ func (c *importChecker) checkIndexBuildingJob(job ImportJob) {
...
+	log := log.With(zap.Int64("jobID", job.GetJobID()))
+	if !job.GetAutoCommit() {
+		// Wait for explicit CommitImport from the replication platform.
+		return
+	}
+	// auto_commit=true: trigger commit by broadcasting the WAL message.
+	if c.commitImportFn == nil {
+		panic("commitImportFn is nil but auto_commit=true; this is a programming error")
```
Comment (czs007) [inline internal/datacoord/import_checker.go]: nit (non-blocking): this `panic` runs inside the ImportChecker ticker goroutine, so hitting it would crash DataCoord. In production `NewImportChecker` always passes `s.broadcastCommitImportMessage` (only tests pass nil), so it is unreachable in practice — but since it is classified as a programming error, `log.Error` + `return` is safer than a hard panic in a long-running loop.

File: internal/distributed/proxy/httpserver/handler_v2.go [hunk 27]
```diff
@@ -3171,6 +3177,30 @@ func (h *HandlersV2) refreshExternalCollection(ctx context.Context, c *gin.Conte
...
+		HTTPAbortReturn(c, http.StatusOK, gin.H{HTTPReturnCode: merr.Code(paramErr), HTTPReturnMessage: paramErr.Error()})
+		return nil, paramErr
+	}
+	req := &datapb.CommitImportRequest{
+		Base:  commonpbutil.NewMsgBase(),
+		JobId: jobID,
+	}
+	c.Set(ContextRequest, req)
+	resp, err := wrapperProxy(ctx, c, req, false, false,
+		"/milvus.proto.milvus.MilvusService/CommitImport",
```
Comment (czs007) [inline internal/distributed/proxy/httpserver/handler_v2.go]: nit (non-blocking): the fullMethod string `/milvus.proto.milvus.MilvusService/CommitImport` (and `/AbortImport` at line 3235) does not correspond to a real MilvusService gRPC method — these are DataCoord-internal RPCs. `checkAuth=false` so no privilege lookup happens, but it still flows into trace/metric labels and contradicts the data_coord.proto comment "internal only, not exposed in public MilvusService". Suggest using a recognizable internal identifier instead.

File: internal/datacoord/import_meta.go [hunk 28]
```diff
@@ -335,3 +338,36 @@ func (m *importMeta) TaskStatsJSON(ctx context.Context) string {
...
+	if job == nil {
+		return merr.WrapErrImportFailed(fmt.Sprintf("job %d not found", jobID))
+	}
+	// Idempotency: if vchannel already committed, skip.
+	for _, c := range job.GetCommittedVchannels() {
+		if c == vchannel {
+			return nil
+		}
+	}
+	// Callback first, persist second.
```
Comment (czs007) [inline internal/datacoord/import_meta.go:356]: (non-blocking) The callback-first/persist-second ordering is well-reasoned and the idempotency argument is sound. One semantic worth documenting explicitly: segment meta (`is_importing=false`) is persisted before the job meta `committed_vchannels`, so a vchannel's data can be visible while the job is still `Committing` (not yet `Completed`). This is inherent to per-vchannel commit fences and is acceptable, but it should be stated in the MEP / a code comment so downstream readers do not assume all-or-nothing visibility.

File: internal/datacoord/import_checker.go (hunk 22 above)
Comment (bigsheeper) [reply internal/datacoord/import_checker.go:462]: Fixed in 2f33b08.

checkIndexBuildingJob now transitions to Uncommitted after index building instead of clearing is_importing and completing the job directly. Segment visibility is now only changed by HandleCommitVchannel after the per-vchannel CommitImport WAL fence is consumed.

The existing 2PC integration tests should now exercise the real non-empty path because imports can actually reach Uncommitted.

File: internal/datacoord/services.go (hunk 23 above)
Comment (bigsheeper) [reply internal/datacoord/services.go]: Fixed in 2f33b08.

broadcastCommitImportMessage and broadcastRollbackImportMessage now use WithBroadcast(job.GetVchannels()) instead of the control channel, so the WAL flusher handles the message on real data vchannels and calls HandleCommitVchannel.

I also added unit coverage that captures the broadcast message and verifies the target channels are the job data vchannels.

File: internal/datacoord/ddl_callbacks_import.go (hunk 24 above)
Comment (bigsheeper) [reply internal/datacoord/ddl_callbacks_import.go:202]: Addressed the code-side part in 2f33b08.

I kept the resource-key-lock approach instead of restoring CAS because these callbacks are serialized by the broadcaster exclusive collection-level resource key. I also added TestImportAckCallbacks_CommitVsAbort_Race to cover the commit-vs-abort invariant.

I will update the MEP separately to replace the old CAS wording with the resource-key-lock concurrency guarantee.

File: internal/datacoord/ddl_callbacks_import.go (hunk 25 above)
Comment (bigsheeper) [reply internal/datacoord/ddl_callbacks_import.go:255]: Confirmed and clarified in tests.

rollbackImportV2AckCallback remains minimal and only transitions the job to Failed. RequestedDiskSize release and CleanupTs setup are already handled by UpdateJobState(Failed), while segment drop stays in the existing import inspector failed path so we keep a single cleanup path.

I added assertions that Failed releases RequestedDiskSize and sets CleanupTs. I will update the MEP to reflect this responsibility split.

File: internal/datacoord/import_checker.go (hunk 26 above)
Comment (bigsheeper) [reply internal/datacoord/import_checker.go]: Fixed in 2f33b08.

checkUncommittedJob now logs an error and returns when commitImportFn is nil with auto_commit=true, instead of panicking from the ticker goroutine.

File: internal/datacoord/import_meta.go (hunk 28 above)
Comment (bigsheeper) [reply internal/datacoord/import_meta.go:356]: Added a code comment in 2f33b08.

HandleCommitVchannel now documents that segment meta is persisted as visible before job meta records committed_vchannels, so a vchannel's imported data may be visible while the job is still Committing. This is intentional per-vchannel fence semantics, not job-level all-or-nothing visibility.

I will mirror this clarification in the MEP update.

File: internal/distributed/proxy/httpserver/handler_v2.go (hunk 27 above)
Comment (bigsheeper) [reply internal/distributed/proxy/httpserver/handler_v2.go]: Fixed in 2f33b08.

The REST handler now uses the internal DataCoord identifiers:
- /milvus.proto.data.DataCoord/CommitImport
- /milvus.proto.data.DataCoord/AbortImport

These are trace/metric labels only; CommitImport/AbortImport are still DataCoord-internal RPCs.

File: internal/datacoord/import_checker.go (hunk 22 above)
Comment (czs007) [reply internal/datacoord/import_checker.go:462]: Verified at HEAD 7d824d1.

checkIndexBuildingJob now hands off to Uncommitted regardless of auto_commit, and unsetSegmentImporting has been removed entirely (0 production callers across the repo — only test comments remain). Segment visibility is gated solely by HandleCommitVchannel after the per-vchannel WAL commit fence is consumed. State machine matches MEP #29. Thanks for the fix.

File: internal/datacoord/services.go (hunk 23 above)
Comment (czs007) [reply internal/datacoord/services.go]: Verified at HEAD 7d824d1.

broadcastCommitImportMessage / broadcastRollbackImportMessage both use WithBroadcast(job.GetVchannels()) now; the flusher dispatches on real data vchannels and HandleCommitVchannel is called per vchannel. The 7d824d1 follow-up that downgrades ErrChannelNotFound from panic to log+return is reasonable in spirit, but has a corner-case semantic worth tightening — posted as a separate non-blocking nit.

File: tests/integration/import/2pc_import_test.go [hunk 29]
```diff
@@ -0,0 +1,427 @@
...
+	err = WaitForImportState(ctx, s.Cluster, jobID, internalpb.ImportJobState_Completed)
+	s.NoError(err)
+
+	// Query AFTER commit - data should be visible
+	countAfter := s.queryRowCount(ctx, collectionName)
+	s.Equal(twoPCRowCount, countAfter, "data should be visible after commit")
+	log.Info("verified data visible after commit", zap.Int("count", countAfter))
+}
+
+func TestTwoPCImport(t *testing.T) {
```
Comment (czs007) [inline tests/integration/import/2pc_import_test.go:425]: (non-blocking) Across the three CI runs since the fix (b77cefb / 2f33b08 / 7d824d1) the ut-integration "Int passed" count has stayed at 46 (UT-GO went 1008→1012, matching the new unit tests, but the integration count did not move). The file's package declaration is `importv2` (same as siblings), TestTwoPCImport is a top-level Test* function with no build tag or `t.Skip`, and scripts/run_intergration_test.sh does `go list -tags dynamic,test ./...` so it should be discovered.

Could you confirm where TestTwoPCImport actually executes end-to-end (ut-integration vs e2e-default vs local), and ideally paste a local `go test -tags dynamic,test ./tests/integration/import/ -run TestTwoPCImport -v` log? `ci-v2/e2e-default` is currently failing across multiple reruns, so the public CI signal does not yet confirm the 2PC end-to-end path.

File: internal/streamingnode/server/flusher/flusherimpl/wal_flusher.go [hunk 30]
```diff
@@ -280,6 +283,49 @@ func (impl *WALFlusherImpl) dispatch(msg message.ImmutableMessage) (err error) {
...
+			impl.logger.DPanic("failed to parse CommitImportMessage", zap.Error(err))
+			return nil
+		}
+		vchannel := msg.VChannel()
+		jobID := commitMsg.Header().GetJobId()
+
+		// Flush DML data before this commit fence. Panic on failure so WAL replays the message.
+		if err := resource.Resource().WriteBufferManager().
+			FlushChannel(context.Background(), vchannel, msg.TimeTick()); err != nil {
+			if errors.Is(err, merr.ErrChannelNotFound) {
```
Comment (czs007) [inline internal/streamingnode/server/flusher/flusherimpl/wal_flusher.go:298]: (non-blocking) When ErrChannelNotFound is hit and we return nil here, HandleCommitVchannel is never called for this vchannel → committed_vchannels never includes it → checkCommittingJob's `len(committed) < len(Vchannels)` stays true for this job forever, until tryTimeoutJob expires it on TimeoutTs (commonly hours).

Functionally it does not deadlock, but it leaves the job stuck in `Committing` for a long time, which pollutes GetImportProgress display and the replication platform's poll loop. Suggest proactively transitioning the job to Failed here (e.g. a fast-fail RPC to DataCoord with reason "vchannel dropped"), rather than relying on the timeout fallback.

Comment (bigsheeper) [reply internal/streamingnode/server/flusher/flusherimpl/wal_flusher.go:298]: Fixed in a94eba65f8. ErrChannelNotFound now only skips the local FlushChannel call and continues to HandleCommitVchannel, so DataCoord still records the vchannel in committed_vchannels and the job can leave Committing. Unexpected FlushChannel errors still panic for WAL replay.

File: tests/integration/import/2pc_import_test.go (hunk 29 above)
Comment (bigsheeper) [reply tests/integration/import/2pc_import_test.go:425]: I checked the discovery path. `scripts/run_intergration_test.sh` runs from `tests/integration` and enumerates packages with `go list -tags dynamic,test ./...`; locally that includes the import package with `2pc_import_test.go` in TestGoFiles:

```text
github.com/milvus-io/milvus/tests/integration/import importv2 ... [2pc_import_test.go ...]
```

The `DONE 46 tests` number staying constant across b77cefb / 2f33b08 / 7d824d1 is not a signal that this file was skipped: `TestTwoPCImport` was already added earlier in the PR history (0010edccb5), so those later fix commits did not add another integration test. CI build 11793 also did run the import package (`Running test cases at github.com/milvus-io/milvus/tests/integration/import ...`), but the gotestsum compact output does not print passing subtest names.

I also tried the exact focused local command. It gets past package discovery but my local C++ artifact is stale for this rebased branch and link fails before test execution (`libmilvus_core` is missing `InitArrowReaderConfig` / `SegmentSetCommitTimestamp`). Current CI for a94eba65f8 is running as integration build 11854; I will use that as the authoritative end-to-end signal once it finishes.

**Review by czs007 — APPROVED** [review]
/lgtm
